pipeline/review/main.go

```diff
@@ -87,9 +96,20 @@ func realMain(ctx context.Context) error {
 // getConfigFromFlags returns review.CommitApprovalPipelineConfig struct
 // using flag values. returns an error if any of the flags have malformed
 // data.
-func getConfigFromFlags() (*review.CommitApprovalPipelineConfig, error) {
+func getConfigFromFlags(ctx context.Context) (*review.CommitApprovalPipelineConfig, error) {
+	// only request a token if none was provided via the githubToken flag
 	if githubToken == "" {
-		return nil, fmt.Errorf("a non-empty github-token must be provided")
+		if githubAppID == "" && githubAppInstallationID == "" && githubAppPrivateKeyResourceName == "" {
+			return nil, fmt.Errorf("a non-empty github-token or GitHubApp values must be provided")
+		}
+		// empty githubToken, but non-empty GitHubApp flags, try requesting a token
+		// using the GitHubApp flags
+		token, err := requestToken(ctx)
+		if err != nil {
+			return nil, err
+		}
+		// successfully got the token, set githubToken to the value we got
+		githubToken = token
 	}
```

Comment:
I'm not sure this is the right pattern. 

1. We should fail if a github token and appID are given, since it's not clear what behavior the user wants and it will lead to unexpected behaviors.

1. We should create an interface that vends a GitHub token, and have different implementations for "static" and "app". That will allow us to support other authentication in the future.


```go
type GitHubAuther interface {
  GitHubToken() (string, error)
}

type GitHubAuthToken struct {
  token string
}

func (g *GitHubAuthToken) GitHubToken() (string, error) {
  return g.token, nil
}

type GitHubAuthApp struct {
  githubAppID, githubAppInstallationID, githubAppSecret string
}

func (g *GitHubAuthApp) GitHubToken() (string, error) {
  // do dance
}
```

Reply:
I tried using this suggested interface structure. Let me know if this is what you had in mind or not.

Reply:
I've moved the interface and respective implementations to auth package.